HTML5: Add cache buster for html export

tijmenvangulik opened the pull request:
Html export works fine for the first time. You can use this for publishing your game on a webserver. But when you publish the game for a second time on your website the end users will not get your new game, instead it just loads the old version from the cache (even if you refresh the page) This is often something which you do not notice during development because during debugging the browser is configured to hard refresh your website.  For this problem I have implemented a cache buster. It is a simple change which adds a parameter with a timestamp of the export moment to the javascript and wasm urls. This way the game is updated when you  place a new version on your website. The fix is for normal html exports and not for progressive web app exports

tijmenvangulik committed:
cache buster for html export

Update platform/javascript/export/export_plugin.cpp

Co-authored-by: Max Hilbrunner <mhilbrunner@users.noreply.github.com>

Much cleaner implementation

Remove tab which was inserted by my editor

Should have removed the not needed var

Reverted space changed

Use utc time

Removed more tabs in code

Calinou commented on platform/javascript/js/engine/preloader.js:
```diff
@@ -32,7 +32,8 @@ const Preloader = /** @constructor */ function () { // eslint-disable-line no-un
 			done: false,
 		};
-		return fetch(file).then(function (response) {
+		var addCacheBuster=typeof CACHE_BUSTER=="string"?CACHE_BUSTER:"";
```
Use `const` for a variable that's only assigned once, and use `===` instead of `==`:

```suggestion
		const addCacheBuster = typeof CACHE_BUSTER === "string" ? CACHE_BUSTER : "";
```

Calinou commented on platform/javascript/js/engine/preloader.js:
```diff
@@ -32,7 +32,8 @@ const Preloader = /** @constructor */ function () { // eslint-disable-line no-un
 		};
-		return fetch(file).then(function (response) {
+		var addCacheBuster=typeof CACHE_BUSTER=="string"?CACHE_BUSTER:"";
+		return fetch(file+addCacheBuster).then(function (response) {
```
```suggestion
		return fetch(file + addCacheBuster).then(function (response) {
```

Calinou commented on platform/javascript/export/export_plugin.cpp:
```diff
@@ -142,12 +143,18 @@ void EditorExportPlatformJavaScript::_fix_html(Vector<uint8_t> &p_html, const Re
 	}
-
+	else {
+		String date=Time::get_singleton()->get_date_string_from_system();
+		String time=Time::get_singleton()->get_time_string_from_system();
+		uint64_t microSec= Time::get_singleton()->get_ticks_msec() % 1000;
+		cache_buster = vformat( String::utf8("?v=%sT%s.%d"),date ,time,microSec);
```
Is adding milliseconds really necessary? I doubt you'll export a project more than once per second :slightly_smiling_face: 

Either way, formatting should be fixed and `snake_case` should be used for variable names in C++:

```suggestion
		const String date = Time::get_singleton()->get_date_string_from_system();
		const String time = Time::get_singleton()->get_time_string_from_system();
		const uint64_t ms = Time::get_singleton()->get_ticks_msec() % 1000;
		const cache_buster = vformat(String::utf8("?v=%sT%s.%d"), date, time, ms);
```

Calinou commented on misc/dist/html/full-size.html:
```diff
@@ -133,7 +133,9 @@
 	</div>
-
+	<script>
+		const CACHE_BUSTER="$CACHE_BUSTER";
```
Can this be moved to the inline script entry below (the one that starts with `const GODOT_CONFIG`)?

akien-mga replied:
This wasn't resolved fully.

tijmenvangulik replied:
Oh I forgot to comment. it is very tricky to move the line to the godot_config. The script in between needs it (the line src='$GODOT_URL'>). You can better not declare something after it is used in javascript. This would also make it a bit unclear what is happening.

tijmenvangulik commented:
Thank you for your quick review. I agree with your changes. Yes it will not happen often that a second export is within I second. But these days with automated scripts you never know. It is such a simple low impact change and it makes it a lot saver, so I included it.

akien-mga commented on platform/javascript/export/export_plugin.h:
```diff
@@ -40,6 +40,7 @@
 #include "main/splash.gen.h"
 #include "platform/javascript/logo.gen.h"
 #include "platform/javascript/run_icon.gen.h"
+#include "core/os/time.h"
```
You should include the header where you use it (in the .cpp). Also make sure to follow the style guidelines: https://docs.godotengine.org/en/latest/community/contributing/code_style_guidelines.html#header-includes

akien-mga commented on platform/javascript/export/export_plugin.cpp:
```diff
@@ -142,12 +143,18 @@ void EditorExportPlatformJavaScript::_fix_html(Vector<uint8_t> &p_html, const Re
 				p_name + ".service.worker.js');}});</script>\n";
 	}
-
+	else {
```
Style error, this should be `} else {`. See https://docs.godotengine.org/en/latest/community/contributing/code_style_guidelines.html

akien-mga commented on platform/javascript/export/export_plugin.cpp:
```diff
@@ -142,12 +143,18 @@ void EditorExportPlatformJavaScript::_fix_html(Vector<uint8_t> &p_html, const Re
+	else {
+		const String date = Time::get_singleton()->get_date_string_from_system();
+		const String time = Time::get_singleton()->get_time_string_from_system();
+		const uint64_t ms = Time::get_singleton()->get_ticks_msec() % 1000;
```
That's not the clock time in ms, that's the amount of time since the editor process was started. It's pretty arbitrary so I don't think it's relevant to export.

tijmenvangulik replied:
Thank you for your comment. Yes I agree it is not the clock time in milliseconds. But that does not matter for a cache buster. It does not have to be a valid time. It just needs to be unique within the second. (this is just in case that an automated UI unit tester presses two times on the button within a second and compares the result)

jordo replied:
As an aside, it would be great to have a UUID in godot core...  It's probably useful in a whole bunch of places in the engine.

akien-mga commented on platform/javascript/export/export_plugin.cpp:
```diff
@@ -142,12 +143,18 @@ void EditorExportPlatformJavaScript::_fix_html(Vector<uint8_t> &p_html, const Re
 	}
-
+	else {
+		const String date = Time::get_singleton()->get_date_string_from_system();
+		const String time = Time::get_singleton()->get_time_string_from_system();
```
You can use `get_datetime_string_from_system()` directly.

Might be worth setting `utc` to true too.

akien-mga commented on platform/javascript/export/export_plugin.cpp:
```diff
@@ -142,12 +143,18 @@ void EditorExportPlatformJavaScript::_fix_html(Vector<uint8_t> &p_html, const Re
+		const String date = Time::get_singleton()->get_date_string_from_system();
+		const String time = Time::get_singleton()->get_time_string_from_system();
+		const uint64_t ms = Time::get_singleton()->get_ticks_msec() % 1000;
+		cache_buster = vformat(String::utf8("?v=%sT%s.%d"), date, time, ms);
```
```suggestion
		cache_buster = vformat("?v=%sT%s.%d", date, time, ms);
```

Should work just as well here.

tijmenvangulik commented:
Thank you for your commet. Yes I agree it is not the clock time. But that does not matter for a cache buster. It does not have to be a valid time. It just needs to be unique within the second.

Tijmen

> On 16 Jan 2022, at 19:25, Rémi Verschelde ***@***.***> wrote:
> 
> 
> @akien-mga commented on this pull request.
> 
> In platform/javascript/export/export_plugin.cpp <https://github.com/godotengine/godot/pull/56841#discussion_r785478499>:
> 
> > @@ -142,12 +143,18 @@ void EditorExportPlatformJavaScript::_fix_html(Vector<uint8_t> &p_html, const Re
>  		head_include += "<script type='application/javascript'>window.addEventListener('load', () => {if ('serviceWorker' in navigator) {navigator.serviceWorker.register('" +
>  				p_name + ".service.worker.js');}});</script>\n";
>  	}
> -
> +	else {
> +		const String date = Time::get_singleton()->get_date_string_from_system();
> +		const String time = Time::get_singleton()->get_time_string_from_system();
> +		const uint64_t ms = Time::get_singleton()->get_ticks_msec() % 1000;
> That's not the clock time in ms, that's the amount of time since the editor process was started. It's pretty arbitrary so I don't think it's relevant to export.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/godotengine/godot/pull/56841#pullrequestreview-853706900>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAVG7NSXVVJIC6VVKGZSTK3UWMETFANCNFSM5MC6I5JA>.
> Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>. 
> You are receiving this because you authored the thread.
>

akien-mga commented on platform/javascript/export/export_plugin.cpp:
```diff
@@ -29,6 +29,7 @@
 /*************************************************************************/
 
 #include "export_plugin.h"
+#include "core/os/time.h"
```
This should be:
```suggestion
#include "export_plugin.h"

#include "core/os/time.h"
```
As per https://docs.godotengine.org/en/latest/community/contributing/code_style_guidelines.html#header-includes

akien-mga commented:
> Yes I agree it is not the clock time. But that does not matter for a cache buster. It does not have to be a valid time. It just needs to be unique within the second.

I'm not sure it's correct though, as the only realistic way to have two exports done in the same second is to do it from two separate processes (e.g. two command line exports on the same machine, or exports done from two different machines and ending at the same time). In this case, `get_ticks_msec` will be process specific and there's no guarantee than a bigger value actually identifies an export that is newer, it just means that the editor instance that did this export had more uptime.

tijmenvangulik commented:
When you write an automated UI test and you test the export by clicking on it. It is possible to click two times within one second on the export button and compare the results and since the cache bustor has the same value in when it does not add mili seconds the unit test will fill because it sees the same result.

Tijmen

> On 16 Jan 2022, at 19:42, Rémi Verschelde ***@***.***> wrote:
> 
> 
> Yes I agree it is not the clock time. But that does not matter for a cache buster. It does not have to be a valid time. It just needs to be unique within the second.
> 
> I'm not sure it's correct though, as the only realistic way to have two exports done in the same second is to do it from two separate processes (e.g. two command line exports on the same machine, or exports done from two different machines and ending at the same time). In this case, get_ticks_msec will be process specific and there's no guarantee than a bigger value actually identifies an export that is newer, it just means that the editor instance that did this export had more uptime.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/godotengine/godot/pull/56841#issuecomment-1013929704>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAVG7NTR6IVHVPQENGWGI43UWMGQDANCNFSM5MC6I5JA>.
> Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>. 
> You are receiving this because you authored the thread.
>

akien-mga commented:
I'm not fully convinced that this is meaningful, but ok, it shouldn't hurt. I'd rename the variable to `ticks` though to make it clear that the code author understood that they're handle engine ticks (i.e. multiple seconds, possibly dozens of minutes) and not the milliseconds that would be expected after `date` and `time`.

tijmenvangulik commented:
Good point. Yes tick is a better name

> On 16 Jan 2022, at 19:59, Rémi Verschelde ***@***.***> wrote:
> 
> 
> I'm not fully convinced that this is meaningful, but ok, it shouldn't hurt. I'd rename the variable to ticks though to make it clear that the code author understood that they're handle engine ticks (i.e. multiple seconds, possibly dozens of minutes) and not the milliseconds that would be expected after date and time.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/godotengine/godot/pull/56841#issuecomment-1013932459>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAVG7NWYXYNOM7XSTKHITU3UWMISDANCNFSM5MC6I5JA>.
> Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>. 
> You are receiving this because you authored the thread.
>

tijmenvangulik commented:
I changed the variable name and I have also added a comment in the code to explain the code buster formatting.

> On 16 Jan 2022, at 19:59, Rémi Verschelde ***@***.***> wrote:
> 
> 
> I'm not fully convinced that this is meaningful, but ok, it shouldn't hurt. I'd rename the variable to ticks though to make it clear that the code author understood that they're handle engine ticks (i.e. multiple seconds, possibly dozens of minutes) and not the milliseconds that would be expected after date and time.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/godotengine/godot/pull/56841#issuecomment-1013932459>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAVG7NWYXYNOM7XSTKHITU3UWMISDANCNFSM5MC6I5JA>.
> Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>. 
> You are receiving this because you authored the thread.
>

tijmenvangulik commented:
I have been able to move the CACHE_BUSTER in the javascript file as requested. I have double checked if every thing still works, and it works (It works because the load trigger is initiated by startGame and at that moment the CACHE_BUSTER constant is declared)

So every I have now fixed every thing you asked. Or is there still more?

Tijmen

> On 16 Jan 2022, at 19:29, Rémi Verschelde ***@***.***> wrote:
> 
> 
> @akien-mga commented on this pull request.
> 
> In misc/dist/html/full-size.html:
> 
> > @@ -133,7 +133,9 @@
>  		</div>
>  		<div id='status-notice' class='godot' style='display: none;'></div>
>  	</div>
> -
> +	<script>
> +		const CACHE_BUSTER="$CACHE_BUSTER";
> 
> This wasn't resolved fully.
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android. 
> You are receiving this because you authored the thread.
>

akien-mga commented:
There seems to still be a style issue reported by CI (trailing whitespace probably): https://github.com/godotengine/godot/runs/4844786586?check_suite_focus=true

Otherwise squashing the commits would indeed be welcome, you can do so as described in https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#the-interactive-rebase

I was waiting for @Faless (HTML5 platform maintainer) to review before suggesting you do more work, but you're welcome to do it now so that it's ready to merge if approved.

tijmenvangulik commented:
I squashed it to one commit. It does not yet run the automated tests again. But maybe I have to wait.

> On 18 Jan 2022, at 20:16, Rémi Verschelde ***@***.***> wrote:
> 
> 
> There seems to still be a style issue reported by CI (trailing whitespace probably): https://github.com/godotengine/godot/runs/4844786586?check_suite_focus=true <https://github.com/godotengine/godot/runs/4844786586?check_suite_focus=true>
> Otherwise squashing the commits would indeed be welcome, you can do so as described in https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#the-interactive-rebase <https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#the-interactive-rebase>
> I was waiting for @Faless <https://github.com/Faless> (HTML5 platform maintainer) to review before suggesting you do more work, but you're welcome to do it now so that it's ready to merge if approved.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/godotengine/godot/pull/56841#issuecomment-1015742206>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAVG7NU72XIMLNL3XQN63LDUWW4CFANCNFSM5MC6I5JA>.
> Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>. 
> You are receiving this because you authored the thread.
>

Faless commented:
Sorry, I didn't have the time to fully investigate this yet, but I don't think we need it, at least not this way.
If your server is properly configured (via E-Tag, or "must-revalidate") you don't need a cache buster, additionally, this might even force re-downloading of some assets that hasn't changed.
If anything, we seem to have to opposite problem, i.e. that the PWA app does not cache enough (#56103 and there we might need to invalid cache see https://github.com/godotengine/godot/compare/3.x...Faless:js/3.x_pwa_prefer_cache which is still WIP )

Calinou commented:
> If your server is properly configured (via E-Tag, or "must-revalidate") you don't need a cache buster, additionally, this might even force re-downloading of some assets that hasn't changed.

Unfortunately, those headers can't be overridden on itch.io or GitHub Pages.

mhilbrunner commented on platform/javascript/export/export_plugin.cpp:
```diff
@@ -141,13 +144,20 @@ void EditorExportPlatformJavaScript::_fix_html(Vector<uint8_t> &p_html, const Re
+		const String date = Time::get_singleton()->get_date_string_from_system();
+		const String time = Time::get_singleton()->get_time_string_from_system();
+		const uint64_t ticks = Time::get_singleton()->get_ticks_msec() % 1000;
+		//the cache buster is a unique value for the export. The value looks like a timestamp but due to
+		//the last value is read from the ticks so it is not an exact time stamp
```
```suggestion
		// The cache buster is a unique value for the export.
		// The value may look like a timestamp but due to the last value being ticks it is not.
```

tijmenvangulik commented:
Yes I have also the problem that I can not override these headers, and you do not want to browser re-evaluate each time the page is loaded (costs to many requests), The only way a browser can see without a server request that a resources is changed is by changing it (so add a cache buster). And yes this can result in loading resources which where not changed because, this will happen not too often and is not in comparison with browser which always needs to check if resources are changed, this will delay loading the page.

Tijmen


> On 19 Jan 2022, at 01:31, Hugo Locurcio ***@***.***> wrote:
> 
> 
> If your server is properly configured (via E-Tag, or "must-revalidate") you don't need a cache buster, additionally, this might even force re-downloading of some assets that hasn't changed.
> 
> Unfortunately, those headers can't be overridden for all files on itch.io or GitHub Pages.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/godotengine/godot/pull/56841#issuecomment-1015959596>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAVG7NTJRZNQZJ6CUPOEKR3UWYA6PANCNFSM5MC6I5JA>.
> Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>. 
> You are receiving this because you authored the thread.
>

tijmenvangulik commented:
The best solution would be not to use one date time as cache buster but to determine the hash of each file and add the hash to the url. This way the file only gets downloaded when it is actually changed. This would be more work, and if we choose to do this a more experienced godot developer should pick this up. However a timestamp is small simple and will work just fine as short term solution which can be back ported to the 3.x because it is a small change.

Tijmen

> On 19 Jan 2022, at 01:31, Hugo Locurcio ***@***.***> wrote:
> 
> 
> If your server is properly configured (via E-Tag, or "must-revalidate") you don't need a cache buster, additionally, this might even force re-downloading of some assets that hasn't changed.
> 
> Unfortunately, those headers can't be overridden for all files on itch.io or GitHub Pages.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/godotengine/godot/pull/56841#issuecomment-1015959596>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAVG7NTJRZNQZJ6CUPOEKR3UWYA6PANCNFSM5MC6I5JA>.
> Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>. 
> You are receiving this because you authored the thread.
>

tijmenvangulik commented:
It is appended for the wasm in the javascript script file

	return fetch(file + cacheBuster).then(function (response) {

When you are going to do effort to add guid you may also consider to use file hashes in this case. The advantage of this is that when the file file is not changed the hash is also not changed.

But I gave up on this pull request, I know the code works fine but it just takes too long to get it merged into the main repository. I have created my own work around to get it fixed.

Tijmen

> On 9 Feb 2022, at 17:08, Jordan Schidlowsky ***@***.***> wrote:
> 
> 
> I'm also a bit confused... In the PR I see the cache buster extension to the .js filename, but i don't see it appended anywhere else. Where is the filename changes to exported .pck and .wasm? To be fair I haven't tried it yet, but from the PR I cant where or tell if this happens
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/godotengine/godot/pull/56841#issuecomment-1033927845>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAVG7NUXZOZKHYPXJUW6B5LU2KGQLANCNFSM5MC6I5JA>.
> Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>. 
> You are receiving this because you authored the thread.
>

jordo commented:
`If the cache headers from the server are wrong, the html page returned will be the old one, with the old hash, and cache will not be busted.`  This is exactly the problem this would be addressing.  We currently have this issue, and this sucks for cached clients because our game is multiplayer and our servers will not let old clients connect due to need to update.  

And I can confirm that unless the web request chain receives a very short expires time with 'must-validate' in cache-control header (which would revalidate and receive a 304), there is going to be problem.  This is also the case for our cluster's reverse proxy, as caching may be done at the proxy as well.  

Deployments aren't just limited to itch.io or GitHub Pages as some html5 gaming portals do not implement caching properly.  This is a partial reason why we are attempting to serve our app in an iframe on portals if it's an option, as we have more control there.

Servers can respond with wrong cache headers, which may or may not be out of your control and different exported URIs may help in those scenarios.  I'm not saying this really needs this to be merged, but there is a use-case here.

Faless commented:
> This is exactly the problem this would be addressing.

As I explained, I don't see how this PR addresses this problem, since if the cache headers are wrong, the `html` returned will have the old timestamp, making the whole process pointless.

jordo commented:
> > This is exactly the problem this would be addressing.
> 
> As I explained, I don't see how this PR addresses this problem, since if the cache headers are wrong, the `html` returned will have the old timestamp, making the whole process pointless.

Which is why I asked about the cache buster on the *.wasm and *.pck exports.    Here's a very real scenario that is happening right now:

*.html files served from a webserver with correct cache control (revalidates), while *.pck (unknown file extension to most webservers) are served with say "Cache-control: max-age=604800"  What happens in this case when re-exported projects now have a different .pck file with the same URI?  The web client in this case doesn't even make the request for a new pck, and just plucks it out of it's local cache.

jordo commented:
In general, the response to this PR seems to be that `as long as http servers are configured properly, and reverse proxies and clients are doing what they are supposed to, this PR is uncessary`.  But that is often not the case in many real-world scenarios.  We have seen de-sync plenty of times among our exported game specifically when *.pck resources are plucked from http client's local cache while the other http resources are re-validated and potentially updated.  A guaranteed different exported URI for each resource imho is a pretty good way to bust this up and guarantee client reload on re-export.  So I do support this although there are a couple small issues with the implementation.

Faless commented:
> In general, the response to this PR seems to be that `as long as http servers are configured properly, and reverse proxies and clients are doing what they are supposed to, this PR is uncessary`.

I think there is a misunderstanding here. Adding a cache buster when the servers are configured correctly **is detrimental**. This is the main reason why I am against this PR.

>  But that is often not the case in many real-world scenarios. We have seen de-sync plenty of times among our exported game specifically when *.pck resources are plucked from http client's local cache while the other http resources are re-validated and potentially updated. 

If you feel this can help a common scenario, I think we can add an **optional** cache buster, but I would still recommend fixing the proxy and servers configurations (again, the goal of CDNs and proxy caching is to provide proper caching headers, if that doesn't happen, something is wrong in the infrastructure).

And again, I still don't think this will fix it in the end, and it will just trigger other problems.

So, in case we want this:
- Needs a rebase
- The cache buster should be added as a [config option](https://github.com/godotengine/godot/blob/master/platform/javascript/js/engine/config.js#L109-L115) and not as a separate function.
- Should also be correctly added to the [localFile](https://github.com/godotengine/godot/blob/master/platform/javascript/js/engine/config.js#L294-L307) function.
- Must be added as an export option (off by default).

tijmenvangulik commented:
Good point. I agree it should be optional.

> On 9 Feb 2022, at 21:01, Fabio Alessandrelli ***@***.***> wrote:
> 
> 
> In general, the response to this PR seems to be that as long as http servers are configured properly, and reverse proxies and clients are doing what they are supposed to, this PR is uncessary.
> 
> I think there is a misunderstanding here. Adding a cache buster when the servers are configured correctly is detrimental. This is the main reason why I am against this PR.
> 
> But that is often not the case in many real-world scenarios. We have seen de-sync plenty of times among our exported game specifically when *.pck resources are plucked from http client's local cache while the other http resources are re-validated and potentially updated.
> 
> If you feel this can help a common scenario, I think we can add an optional cache buster, but I would still recommend fixing the proxy and servers configurations (again, the goal of CDNs and proxy caching is to provide proper caching headers, if that doesn't happen, something is wrong in the infrastructure).
> 
> And again, I still don't think this will fix it in the end, and it will just trigger other problems.
> 
> So, in case we want this:
> 
> Needs a rebase
> The cache buster should be added as a config option <https://github.com/godotengine/godot/blob/master/platform/javascript/js/engine/config.js#L109-L115> and not as a separate function.
> Should also be correctly added to the localFile <https://github.com/godotengine/godot/blob/master/platform/javascript/js/engine/config.js#L294-L307> function.
> Must be added as an export option (off by default).
> —
> Reply to this email directly, view it on GitHub <https://github.com/godotengine/godot/pull/56841#issuecomment-1034142583>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAVG7NULLLW3Y7CJDKSYGSLU2LBY5ANCNFSM5MC6I5JA>.
> Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>. 
> You are receiving this because you authored the thread.
>

jordo commented:
If you append the file contents hash to the exported filename (with the exception of the .html  root)  I'm not sure how this can be considered detrimental as you can actually be **more** aggressive on your webserver cache settings...  Here is an example output from say our current export:
```
index.html
index.audio.worklet.js
index.js
index.pck
index.wasm
```

If you want this merged what I would propose is output like:
```
index.html
index-FILECONTENTSHASH.audio.worklet.js
index-FILECONTENTSHASH.js
index-FILECONTENTSHASH.pck
index-FILECONTENTSHASH.wasm
```

Since all the other URI's are loaded via the first request to index.html, *.html is the only resource type that needs to be configured to revalidate.  All other files under this directory you could even cache indefinitely without revalidation as it's guaranteed their contents will never change (and if their content does, it is exported as a different URI).  

From a caching perspective, this is actually the opposite of detrimental and more cache efficient because right now this is what happens if configured correctly:

<img width="707" alt="Screen Shot 2022-02-10 at 2 52 52 PM" src="https://user-images.githubusercontent.com/323868/153494519-06e6e485-5d59-44fe-bc62-4cfe436189e4.png">

Every URI needs to be re-validated because the content MAY change at any time, which mean a bunch of additional round-trips and a bunch of 304's **if** configured properly ( and really annoying de-syncs across wasm, js, and pck files if not).  


If you implement exports like this:
```
index.html
index-FILECONTENTSHASH.audio.worklet.js
index-FILECONTENTSHASH.js
index-FILECONTENTSHASH.pck
index-FILECONTENTSHASH.wasm
```
you could configure your webserver to aggressively instruct clients to cache .js, .wasm, and .pck as their contents at their respective URI's are guaranteed to never change.  In this scenario there is only a single request and single validation +304  response required for the initial request to index.html, as every other dependency is derived from that first request.

At any rate, from a potential cache 'getting it wrong' perspective, I think the above export example is less error prone.  I think it's going to be a good option for the engine to provide devs because it really removes a lot of potential issues whether it's considered a cache buster or not really.  Definitely useful for devs who are deploying to environments that aren't under their direct control, or devs who are not familiar with what their actual server cache settings should be.  However, I think in order for this to make it into main, I would want to the see a file-contents hash on the exported filename though which is definitely more work than what is included in this PR.

The only problem / downside I can see to implementing this is local filesystem issues & confusion if you choose to export to the same output directory as a previous export.  In which case I think there could/would be multiple copies of the same dependency with different hash depending on how implemented, but I can lready think of a few different solutions for that.

Faless commented:
I've honestly never seen cache busting by changing file name, I would definitely go for using a query string, though using the hash instead of timestamp sounds good.

Faless commented:
> If it's implemented with file contents hash instead of a timestamp, this would address your concerns about being 'detrimental' correct?

Yes, I would still make it optional or at least off in the PWA export since it's irrelevant there.

YuriSizov commented:
Seems like some consensus was reached, but this was never finished. Would be nice to revive this for a future 4.x release.

PeterMarques commented:
May be the fix for https://github.com/godotengine/godot/issues/89455 and https://github.com/godotengine/godot/issues/69414.

This needs to be pushed up, and 3.x cherry picked.

tijmenvangulik commented:
Hi,

Cache buster fix the caching problem for normal websites. For PWA’s you need more. I do not use the PWA version because the cache problem is even worse. You will also need to correct the pwa service worker for this which is in my opinion not correctly written for on and offline caching. I have given up on this change. I have simply modified the html template with extra cache busters and wrote a script to update it on every release.

Tijmen

> On 19 Mar 2024, at 03:46, PeterMarques ***@***.***> wrote:
> 
> 
> May be the fix for #89455 <https://github.com/godotengine/godot/issues/89455> and #69414 <https://github.com/godotengine/godot/issues/69414>.
> 
> This needs to be pushed up, and 3.x cherry picked.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/godotengine/godot/pull/56841#issuecomment-2005647522>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAVG7NRGY42OOJLJ2AXQDMTYY6RGTAVCNFSM5MC6I5JKU5DIOJSWCZC7NNSXTN2JONZXKZKDN5WW2ZLOOQ5TEMBQGU3DINZVGIZA>.
> You are receiving this because you authored the thread.
>
